src/source/source_state.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+import { extend } from '../util/util';
+import Tile from './tile';
+
+export type FeatureStates = {[feature_id: string]: {[key: string]: any }};
+export type LayerFeatureStates = {[layer: string]: FeatureStates};
+
+/** 
+ * SourceFeatureState manages the state and state changes
+ * to features in a source, separated by source layer.
+ *
+ * @private
+*/
+class SourceFeatureState {
+    state: LayerFeatureStates;
+    stateChanges: LayerFeatureStates;
+
+    constructor() {
+        this.state = {};
+        this.stateChanges = {};
+    }
+
+    setState(sourceLayer: string, feature: string, key: string, value: any) {
+        feature = String(feature);
+        this.stateChanges[sourceLayer] = this.stateChanges[sourceLayer] || {};
+        this.stateChanges[sourceLayer][feature] = this.stateChanges[sourceLayer][feature] || {};
+        this.stateChanges[sourceLayer][feature][key] = value;
+    }
+
+    getState(sourceLayer: string, feature: string, key?: string) {
+        feature = String(feature);
+        const base = this.state[sourceLayer] || {};
+        const changes = this.stateChanges[sourceLayer] || {};
+
+        if (!key) {
+            return extend({}, base[feature], changes[feature]);
+        }
+        if (changes[feature]) {
+            return changes[feature][key];
+        }
+        if (base[feature]) {
+            return base[feature][key];
+        }
+    }
+
+    initializeTileState(tile: Tile) {
+        tile.updateFeatureState(this.state);
+    }
+
+    coalesceChanges(tiles: {[any]: Tile}) {
+        const changes: LayerFeatureStates = {};
+        for (const sourceLayer in this.stateChanges) {
+            this.state[sourceLayer]  = this.state[sourceLayer] || {};
+            const layerStates = {};
+            for (const id in this.stateChanges[sourceLayer]) {
+                this.state[sourceLayer][id] = extend(
+                                                {},
```

Comment:
Is it necessary to create and extend a new object? or could it just extend the existing `this.state[sourceLayer][id]`?